linkerd/app/core/src/retry.rs

```diff
@@ -183,9 +134,20 @@ where
         &self,
         req: http::Request<A>,
     ) -> Either<Self::RetryRequest, http::Request<A>> {
-        if self.can_retry(&req) {
-            return Either::A(req.map(|body| ReplayBody::new(body, MAX_BUFFERED_BYTES)));
-        }
-        Either::B(req)
+        let (head, body) = req.into_parts();
+        let replay_body = match ReplayBody::try_new(body, MAX_BUFFERED_BYTES) {
```

Comment:
this is really not important at all, but i _kind of_ feel like the `prepare_request` method should be responsible for checking the `size_hint` and `ReplayBody` should just have an unconditional `new` constructor...that kind of feels like a better separation of concerns to me, with the retry policy code being responsible for determining if the body is too large to retry, and `ReplayBody` just providing the buffering mechanism.

but, since `ReplayBody` _does_ still need to be aware of the body size limit so that it can stop buffering if necessary, maybe this is fine...either way, it's not a big deal...

Reply:
I initially had it the way you suggest, but I think I liked to make it impossible to create a `ReplayBody` that is 'illegal' -- this means that the redundant check in `clone_request` becomes totally unnecessary: by virtue of having constructed a `ReplayBody`, we know that it doesn't violate our expectations on body size.